packages/test/test-end-to-end-tests/src/test/testApi.ts

```diff
@@ -0,0 +1,240 @@
+            // assume it is valid if it exists
+            return { version, modulePath };
+        }
+        try {
+            console.log(`Installing version ${version}`);
+
+            // Create the directory
+            mkdirSync(modulePath, { recursive: true });
+
+            // Release the __dirname but lock the modulePath so we can do parallel installs
+            const release2 = await lock(modulePath, { retries: { forever: true } });
+            release();
+            release = release2;
+
+            // Install the packages
+            await new Promise<void>((res, rej) =>
+                exec(`npm init --yes`, { cwd: modulePath }, (error, stdout, stderr) => {
+                    if (error) {
+                        rej(new Error(`Failed to initialize install directory ${modulePath}`));
+                    }
+                    res();
+                }),
+            );
+            await new Promise<void>((res, rej) =>
+                exec(
+                    `npm i --no-package-lock ${packageList.map((pkg) => `${pkg}@${version}`).join(" ")}`,
+                    { cwd: modulePath },
+                    (error, stdout, stderr) => {
+                        if (error) {
+                            rej(new Error(`Failed to install in ${modulePath}`));
+                        }
+                        res();
+                    },
+                ),
+            );
+        } catch (e) {
+            try { unlinkSync(modulePath); } catch (ex) { }
+            throw new Error(`Unable to install version ${version}\n${e}`);
+        }
+        return { version, modulePath };
+    } finally {
+        release();
+    }
+}
+
+export const mochaGlobalSetup = async () => Promise.all(
+    [ensureInstalled(getRequestedRange(-1)), ensureInstalled(getRequestedRange(-2))],
+);
+
+function checkInstalled(requested: string) {
+    const version = resolveVersion(requested);
+    const modulePath = path.join(__dirname, "..", "..", "node_modules", ".legacy", version);
+    if (existsSync(modulePath)) {
+        // assume it is valid if it exists
+        return { version, modulePath };
+    }
+    throw new Error(`Requested version ${requested} resolved to ${version} is not installed`);
+}
+
+const loadPackage = (modulePath: string, pkg: string) =>
+    // eslint-disable-next-line @typescript-eslint/no-unsafe-return, @typescript-eslint/no-require-imports
+    require(path.join(modulePath, "node_modules", pkg));
+
+function getRequestedRange(requested?: number | string): string {
+    if (requested === undefined) { return pkgVersion; }
+    if (typeof requested === "string") { return requested; }
+    const version = new semver.SemVer(pkgVersion);
+    // ask for prerelease in case we just bumpped the version and haven't release the previous version yet.
+    return `^${version.major}.${version.minor + requested}.0-0`;
+}
+
+export function getLoaderApi(requested?: number | string): typeof LoaderApi {
```

Comment:
i was envisioning this to be quite different. i was thinking we have some api, like getLoader, which took an ITestDriver. The method would be defined in test definitions, and implemented in a test loader package, which had a mocha hook. this package would encode creating it's version of the loader from the test driver interface. we could then grab this package from any version and inject it via mocha hook into out tests. no need for dynamic loading

Reply:
I am not 100% wedded to this per se.  But I like this because it can be on the path to self contain (no extra setup and script is necessary), to run conpat versions.  

Also I didn't want to add another level of "compatibility" require for the test API.  The only "compat" are the actual classes.    Just like TestObjectProvider, we should always use the newest version of test code so that you can update it to match the test patterns.  Also, our compat boundary may evolve (i.e. the DataObject <-> DDS) and we want to have the ablility to change that quickly.

Reply:
Additionally, this method allow you to load different version of the same library.  This allow you to write test that test compat between different versions of merge tree, for example.